### DISCUSSION TITLE
Few questions about vlang (not found in docs)

### DISCUSSION THREAD

**Discussion by fairking** [vlang/v#16433] [General]
Hi guys,

I am trying to play with some proof of concept before I can start a small project.

I gathered few questions mostly related to reflection. Maybe someone can help me with that. Please consider to improve the docs if you think it is necessary.

1. Is there any way to declare an optional variable. If not, what is the workaround? (maybe array?). Eg:
``` vlang
struct Foo { }

struct Bar {
    foo ?Foo
}

fn (b Bar) set(f Foo) {
    b.foo = f
}

fn (b Bar) get() ?Foo {
    return b.foo
}
```
Ref: https://github.com/vlang/v/blob/master/doc/docs.md#optionresult-types-and-error-handling

2. How to get a type from T or from the actual instance? If i go [there](https://modules.vlang.io) and search `type` I have millions of things :-( Eg.:
``` vlang
struct Bar {}

struct Foo<T> {
    v T
 }

fn (f Foo) get_type() type {
    return f.v.type // ????
}

// or

fn get_type<T>() type {
    return T.type // ????
}

// or

fn get_type() type {
    b := Bar { }
    return b.type
}

// or

fn get_type() type {
    return Bar.type
}
```
Ref (?): https://github.com/vlang/v/blob/master/doc/docs.md#compile-time-reflection

3. How to add a new element to array? Eg.:
``` vlang
lst := []int
lst.push(1) // ???
}
```
Ref: https://github.com/vlang/v/blob/master/doc/docs.md#arrays

4. Can I declare an array of different structs? Eg.:
``` vlang
struct Foo { }
struct Bar { }

fn getAll() [] {
    lst := [] // ???
    lst.push(Foo{})
    lst.push(Bar{})
    return lst
}
```
Ref: https://github.com/vlang/v/blob/master/doc/docs.md#array-types

5. Please consider to add a `single()` method as well, as I am used to it in other languages. Eg.:
``` vlang
a := [1,2,3]
n := a.filter(it == 2).single() // it will panic if there are none or more than one element
assert n == 2
```

I would be glad to contribute but before I need to learn `v` first 🥇.

Thanks

**Comment by JalonSolov** [comment]
1. You appear to mean a field in a struct, not a simple variable.  The answer is:  No, not yet.  There is, however, an open PR which should make this work.

2. No, at least not yet.  You can compare what you get against specific types, but you cannot get the type directly.

3. Uh... it's right there in the docs, shortly below the link you gave... `An element can be appended to the end of an array using the push operator <<. It can also append an entire array.`

4. No.  You can make an array of sumtypes, but you cannot make an array of different types without using a sumtype.

5. We'll be happy to look at any PR's you submit.  :-)

**Reply by fairking** [reply]
> 4. No. You can make an array of sumtypes, but you cannot make an array of different types without using a sumtype.

I guess it could be something like bellow:
``` vlang
mut a := []&
a << &Foo{}
a << &Bar{}
```

**Reply by JalonSolov** [reply]
Except `[]&` is not a valid syntax.  You can throw type safety, etc., out the windows and just do
```v
mut a := []voidptr
```
but why are you bothering to use a type-safe language if you're just going to do that?
